osmoutils/binary_search.go

```diff
@@ -93,3 +133,41 @@ func BinarySearch(f func(input sdk.Int) (sdk.Int, error),
 	}
 	return curEstimate, nil
 }
+
+// Binary search BigDec inputs between [lowerbound, upperbound] to a monotonic increasing function f
+// We stop once f(found_input) meets the ErrTolerance constraints.
+// If we perform more than maxIterations (or equivalently lowerbound = upperbound), we return an error.
+func BinarySearchBigDec(f func(input osmomath.BigDec) (osmomath.BigDec, error),
+	lowerbound osmomath.BigDec,
+	upperbound osmomath.BigDec,
+	targetOutput osmomath.BigDec,
+	errTolerance ErrTolerance,
+	maxIterations int,
+) (osmomath.BigDec, error) {
+	// Setup base case of loop
+	curEstimate := lowerbound.Add(upperbound).Quo(osmomath.NewBigDec(2))
+	curOutput, err := f(curEstimate)
+	if err != nil {
+		return osmomath.BigDec{}, err
+	}
+	curIteration := 0
+	for ; curIteration < maxIterations; curIteration += 1 {
+		compRes := errTolerance.CompareBigDec(curOutput, targetOutput)
+		if compRes > 0 {
+			upperbound = curEstimate
+		} else if compRes < 0 {
+			lowerbound = curEstimate
+		} else {
+			break
+		}
+		curEstimate = lowerbound.Add(upperbound).Quo(osmomath.NewBigDec(2))
+		curOutput, err = f(curEstimate)
+		if err != nil {
+			return osmomath.BigDec{}, err
+		}
+	}
+	if curIteration == maxIterations {
+		return osmomath.BigDec{}, errors.New("hit maximum iterations, did not converge fast enough")
+	}
+	return curEstimate, nil
```

Comment:
If you go with the previous suggestion, you should be able to simply return error here, without doing the iteration check

Reply:
Updated, please see my comment above! 

https://github.com/osmosis-labs/osmosis/pull/2802/files#r975799847